File: docs/examples/cpp/src/quick_start.cc
#include <reduct/client.h>

#include <iostream>

using reduct::IBucket;
using reduct::IClient;
using sec = std::chrono::seconds;

int main() {
    // 1. Create a ReductStore client
    auto client = IClient::Build("http://127.0.0.1:8383");

    // 2. Get or create a bucket with 1Gb quota
    auto [bucket, create_err] = client->GetOrCreateBucket("my-bucket", {
        .quota_type = IBucket::QuotaType::kFifo,
        .quota_size = 1'000'000'000
    });

    if (create_err) {
        std::cerr << "Error: " << create_err;
        return -1;
    }

    // 3. Write some data with timestamps in the 'sensor-1' entry
    IBucket::Time start = IBucket::Time::clock::now();
    [[maybe_unused]] auto write_err =
            bucket->Write("sensor-1", start, [](auto rec) { rec->WriteAll("Record #1"); });
    write_err = bucket->Write("sensor-1", start + sec(1), [](auto rec) { rec->WriteAll("Record #2"); });

    // 4. Query the data by time range
    auto err = bucket->Query("sensor-1", start,  start + sec(2), {}, [](auto&& record) {
        std::cout << "Timestamp: " << record.timestamp.time_since_epoch().count() << std::endl;
        std::cout << "Content Length: " << record.size << std::endl;

        auto [blob, read_err] = record.ReadAll();
        if (!read_err) {
            std::cout << "Read blob: " << blob << std::endl;
        }

        return true;
    });

    if (err) {
        std::cerr << "Error: " << err;
        return -1;
    }

    // 5. End
    return 0;
}

File: docs/examples/java/src/main/java/store/reduct/QuickStart.java
package store.reduct;

import store.reduct.client.ReductClient;
import store.reduct.client.config.ServerProperties;
import store.reduct.model.bucket.Bucket;
import store.reduct.model.bucket.BucketSettings;
import store.reduct.model.bucket.QuotaType;
import store.reduct.model.record.Record;

import java.net.http.HttpClient;
import java.util.Iterator;

public class QuickStart {
    public static void main(String[] args) {
         // 1. Create a ReductStore client
        ServerProperties properties = ServerProperties
                .builder()
                .url("http://localhost:8383")
                .apiToken("my-token")
                .build();

        HttpClient httpClient = HttpClient.newHttpClient();
        ReductClient client = new ReductClient(properties, httpClient);

        // 2. Get or create a bucket with 1Gb quota
        BucketSettings settings = BucketSettings.builder()
                .quotaType(QuotaType.FIFO)
                .quotaSize(1_000_000_000)
                .exists(true)
                .build();
        Bucket bucket = client.createBucket("my-bucket1", settings);

        // 3. Write some data with timestamps in the 'sensor-1' entry
        long timestamp = System.currentTimeMillis() * 1000; // in microseconds
        Record record = Record.builder()
                .timestamp(timestamp)
                .type("application/octet-stream")
                .body("Record #1".getBytes())
                .build();

        bucket.writeRecord("sensor-1", record);
        record = Record.builder()
                .timestamp(timestamp + 1000_000)
                .type("application/json")
                .length(9)
                .body("Record #2".getBytes())
                .build();
        bucket.writeRecord("sensor-1", record);

        // 4. Read the data back
        for (Iterator<Record> it = bucket.query("sensor-1", timestamp, timestamp + 2000_000, 60L); it.hasNext(); ) {
            Record rec = it.next();

            System.out.println("Record timestamp: " + rec.getTimestamp());
            System.out.println("Record size: " + rec.getBody().length);
        }

    }
}